File: Android/shared/src/main/kotlin/shared/dto/enums/HandleInvite.kt
package shared.dto.enums

import kotlinx.serialization.DeserializationStrategy
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonContentPolymorphicSerializer
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.jsonObject
import shared.dto.objects.Customer
import shared.dto.protocols.Responsable

@Serializable(HandleInvite.Serializer::class)
sealed class HandleInvite : Responsable {
    @Serializable
    data class Success(
        val successCustomer: Customer.Responses.Full,
    ) : HandleInvite()

    @Serializable
    data class Request(
        val requestVerify: Customer.Responses.Verify,
    ) : HandleInvite()

    internal object Serializer :
        JsonContentPolymorphicSerializer<HandleInvite>(HandleInvite::class) {
        override fun selectDeserializer(element: JsonElement): DeserializationStrategy<HandleInvite> {
            return when (element.jsonObject.contains("requestVerify")) {
                true -> Request.serializer()
                false -> Success.serializer()
            }
        }
    }
}

File: Node/src/dto/enums/HandleInvite.ts
import { Customer } from "../objects/Customer"

type Success = Customer.Responses.Full
type Request = Customer.Responses.Verify

export enum Types {
    Success = "success",
    Request = "request",
}

export class HandleInvite {
    type: Types
    invite: Success | Request

    constructor(type: Types, invite: Success | Request) {
        this.type = type
        this.invite = invite
    }

    //TODO: проверить
    toJSON() {
        let key: string
        switch (this.type) {
            case Types.Request:
                key = "verify"
            case Types.Success:
                key = "customer"
        }

        return {
            [this.type]: {
                [key]: this.invite
            }
        }
    }

    //TODO: проверить
    static fromJSON(json: any): HandleInvite {
        const type = Object.keys(json).find(key => Object.values(Types).find(value => value === key));
        if (!type) {
            throw new Error('Invalid JSON: no type found');
        }
        const invite = json[type].invite;

        return new HandleInvite(type as Types, invite);
    }
}